File: libs/libresea/arch/x64/syscall.c
#include <resea.h>
#include <resea/ipc.h>
#include <resea/string.h>

struct thread_info *get_thread_info(void) {
    struct thread_info *info;
    __asm__ __volatile__("rdgsbase %0" : "=a"(info));
    return info;
}

static void copy_message(struct message *dst, const struct message *src) {
    uint32_t header = src->header;
    // TODO: Use memcpy.
    dst->header = header;
    dst->from = src->from;
    dst->notification = src->notification;
    dst->payloads.channel = src->payloads.channel;
    dst->payloads.page = src->payloads.page;

    memcpy(&dst->payloads.data, &src->payloads.data,
           INLINE_PAYLOAD_LEN(header));
}

void copy_to_ipc_buffer(const struct message *m) {
    copy_message(get_ipc_buffer(), m);
}

void copy_from_ipc_buffer(struct message *buf) {
    copy_message(buf, get_ipc_buffer());
}

error_t sys_ipc(cid_t ch, uint32_t ops) {
    error_t err;
    __asm__ __volatile__(
        "syscall"
        : "=a"(err)
        : "a"(SYSCALL_IPC | ops), "D"(ch)
        : "%rsi", "%rdx", "%rcx", "%r8", "%r9", "%r10", "%r11", "memory");
    return err;
}

int sys_open(void) {
    int cid_or_err;
    __asm__ __volatile__(
        "syscall"
        : "=a"(cid_or_err)
        : "a"(SYSCALL_OPEN)
        : "%rdi", "%rsi", "%rdx", "%rcx", "%r8", "%r9", "%r10", "%r11");
    return cid_or_err;
}

error_t sys_close(cid_t ch) {
    int cid_or_err;
    __asm__ __volatile__(
        "syscall"
        : "=a"(cid_or_err)
        : "a"(SYSCALL_CLOSE), "D"(ch)
        : "%rsi", "%rdx", "%rcx", "%r8", "%r9", "%r10", "%r11");
    return cid_or_err;
}

error_t sys_link(cid_t ch1, cid_t ch2) {
    error_t err;
    __asm__ __volatile__("syscall"
                         : "=a"(err)
                         : "a"(SYSCALL_LINK), "D"(ch1), "S"(ch2)
                         : "%rdx", "%rcx", "%r8", "%r9", "%r10", "%r11");
    return err;
}

error_t sys_transfer(cid_t src, cid_t dst) {
    error_t err;
    __asm__ __volatile__("syscall"
                         : "=a"(err)
                         : "a"(SYSCALL_TRANSFER), "D"(src), "S"(dst)
                         : "%rdx", "%rcx", "%r8", "%r9", "%r10", "%r11");
    return err;
}

File: libs/resea/arch/x64/syscall.rs
use crate::result::{Error, Result};

const SYSCALL_IPC: u32 = 0;
const SYSCALL_OPEN: u32 = 1;
const SYSCALL_TRANSFER: u32 = 4;
const IPC_SEND: u32 = 1 << 8;
const IPC_RECV: u32 = 1 << 9;
const IPC_NOBLOCK: u32 = 1 << 10;

unsafe fn convert_error(error: i32) -> Result<()> {
    if error < 0 {
        Err(core::mem::transmute::<u8, Error>(-error as u8))
    } else {
        Ok(())
    }
}

unsafe fn ipc_syscall(cid: i32, ops: u32) -> Result<()> {
    let error: i32;
    asm!(
        "syscall"
        : "={eax}"(error)
        : "{eax}"(SYSCALL_IPC | ops),
          "{rdi}"(cid)
        : "memory", "rdi", "rsi", "rdx", "rcx", "r8", "r9", "r10", "r11"
        : "volatile"
    );

    convert_error(error)
}

pub unsafe fn nop() {
    let error: i32;
    asm!(
        "nop"
        : "={eax}"(error)
        : "{eax}"(6),
          "{rdi}"(0xdead)
        : // "rsi", "rdx", "rcx", "r8", "r9", "r10", "r11"
        : "volatile"
    );
}

pub unsafe fn open() -> Result<i32> {
    let cid_or_error: i32;
    asm!(
        "syscall"
        : "={eax}"(cid_or_error)
        : "{eax}"(SYSCALL_OPEN)
        : "rdi", "rsi", "rdx", "rcx", "r8", "r9", "r10", "r11"
        : "volatile"
    );

    if cid_or_error < 0 {
        Err(core::mem::transmute::<u8, Error>(-cid_or_error as u8))
    } else {
        Ok(cid_or_error)
    }
}

pub unsafe fn transfer(src: i32, dst: i32) -> Result<()> {
    let error: i32;
    asm!(
        "syscall"
        : "={eax}"(error)
        : "{eax}"(SYSCALL_TRANSFER),
          "{rdi}"(src),
          "{rsi}"(dst)
        : "rdi", "rsi", "rdx", "rcx", "r8", "r9", "r10", "r11"
        : "volatile"
    );

    convert_error(error)
}

pub unsafe fn send(cid: i32) -> Result<()> {
    ipc_syscall(cid, IPC_SEND)
}

pub unsafe fn send_noblock(cid: i32) -> Result<()> {
    ipc_syscall(cid, IPC_SEND | IPC_NOBLOCK)
}

pub unsafe fn recv(cid: i32) -> Result<()> {
    ipc_syscall(cid, IPC_RECV)
}

pub unsafe fn call(cid: i32) -> Result<()> {
    ipc_syscall(cid, IPC_SEND | IPC_RECV)
}
